### PR TITLE
Unload and load module

### PR DESCRIPTION
**Pull request by SnapDragon7410**

### CODE REVIEW

**Comment by SnapDragon7410** [comment]
@SnapDragon7410 https://github.com/AvinashReddy3108/PaperplaneExtended/pull/30

**Comment by SnapDragon7410** [comment]
@SnapDragon7410 Anybody can review

File: userbot/plug_core.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+async def load(event):
+    if event.fwd_from:
+        return
+    shortname = event.pattern_match["shortname"]
+    try:
+        try:
+            remove_plugin(shortname)
+        except:
+            pass
+        __import__(f"userbot.modules.{shortname}")
```
Comment (penn5) [inline userbot/plug_core.py]: Do not use `__import__` directly, instead use `importlib`

File: userbot/plug_core.py [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        await event.edit(f"Successfully loaded {shortname}")
+    except Exception as e:
+        await event.edit(f"Could not load {shortname} because of the following error.\n{str(e)}")
```
Comment (penn5) [inline userbot/plug_core.py]: Don't stringify exceptions.

File: userbot/events.py [hunk 3]
```diff
@@ -15,11 +15,16 @@
...
-from userbot import bot, BOTLOG_CHATID
+from userbot import bot, BOTLOG_CHATID, LOAD_PLUG
 
 
 def register(**args):
     """ Register a new event. """
+    import inspect
+    stack = inspect.stack()
+    previous_stack_frame = stack[1]
+    file_test = Path(previous_stack_frame.filename)
```
Comment (penn5) [inline userbot/events.py]: Path is not imported or defined

File: userbot/events.py [hunk 4]
```diff
@@ -15,11 +15,16 @@
...
 from telethon import events
 
-from userbot import bot, BOTLOG_CHATID
+from userbot import bot, BOTLOG_CHATID, LOAD_PLUG
 
 
 def register(**args):
     """ Register a new event. """
+    import inspect
+    stack = inspect.stack()
```
Comment (penn5) [inline userbot/events.py]: `inspect.stack()` is a Cython implementation detail. Don't use it.

File: userbot/events.py [hunk 5]
```diff
@@ -51,6 +56,11 @@ def register(**args):
             args['pattern'] = pattern.replace('^.', unsafe_pattern, 1)
 
     def decorator(func):
+        try:
```
Comment (penn5) [inline userbot/events.py]: WTH! That is seriously the wrong way to go about this!

File: userbot/plug_core.py (hunk 1 above)
Comment (SnapDragon7410) [reply userbot/plug_core.py]: Ok

File: userbot/plug_core.py (hunk 2 above)
Comment (SnapDragon7410) [reply userbot/plug_core.py]: Ok

File: userbot/__main__.py [hunk 6]
```diff
@@ -25,6 +25,7 @@
 
 for module_name in ALL_MODULES:
     imported_module = import_module("userbot.modules." + module_name)
+import_module("userbot.plug_core")
```
Comment (penn5) [inline userbot/__main__.py]: Just add it to ALL_MODULES

**Comment by penn5** [comment]
@AvinashReddy3108 suggesting to close this

File: userbot/plug_core.py (hunk 2 above)
Comment (SnapDragon7410) [reply userbot/plug_core.py]: Should i leave e alone or should i remove it?

File: userbot/events.py (hunk 4 above)
Comment (SnapDragon7410) [reply userbot/events.py]: Then how must i get the modules name?

File: userbot/events.py (hunk 3 above)
Comment (SnapDragon7410) [reply userbot/events.py]: 1. Ok fixed

File: userbot/events.py (hunk 5 above)
Comment (SnapDragon7410) [reply userbot/events.py]: Then tell me how

**Comment by SnapDragon7410** [comment]
> @AvinashReddy3108 suggesting to close this

Heyy dude

File: userbot/__main__.py (hunk 6 above)
Comment (SnapDragon7410) [reply userbot/__main__.py]: No

File: userbot/events.py (hunk 5 above)
Comment (SnapDragon7410) [reply userbot/events.py]: Suggest another method

**Comment by raphielscape** [comment]
Putting a bit of bump

Missing License header will also putting you in a issue with Upstream

**Comment by SnapDragon7410** [comment]
Ok

**Comment by SnapDragon7410** [comment]
> Putting a bit of bump
> 
> Missing License header will also putting you in a issue with Upstream

Done

File: userbot/events.py [hunk 7]
```diff
@@ -15,11 +15,16 @@
 
 from telethon import events
 
-from userbot import bot, BOTLOG_CHATID
+from userbot import bot, BOTLOG_CHATID, LOAD_PLUG
 
 
 def register(**args):
     """ Register a new event. """
+    import inspect
```
Comment (penn5) [inline userbot/events.py]: Move this to the top of the file

File: userbot/events.py (hunk 5 above)
Comment (penn5) [reply userbot/events.py]: You can import the module to a temporary namespace and unload each function it exports

File: userbot/__main__.py (hunk 6 above)
Comment (penn5) [reply userbot/__main__.py]: wth? why?

**Comment by penn5** [comment]
Also why is this being merged into sql-extended? It is derived from sql-dirty and should be merged as such

**Comment by raphielscape** [comment]
Also better lint this with pylint according to the Upstream linting guidelines

**Comment by penn5** [comment]
> Also better lint this with pylint according to the Upstream linting guidelines

As its derived from sql-dirty I don't think lint passes on the sql-dirty branch even

File: userbot/events.py [hunk 8]
```diff
@@ -15,13 +15,13 @@
 from pathlib import Path
 
 from telethon import events
+import stack
```
Comment (penn5) [inline userbot/events.py]: WTH its import inspect!

I bet you haven't even tested it.

**Comment by AvinashReddy3108** [comment]
> Also better lint this with pylint according to the Upstream linting guidelines

Did'nt balaji move over to yapf for linting code ?

**Comment by AvinashReddy3108** [comment]
Closing this issue, coz PR's are to be made in the sql-dirty branch only !!

**Comment by raphielscape** [comment]
> > Also better lint this with pylint according to the Upstream linting guidelines
> 
> Did'nt balaji move over to yapf for linting code ?

Devcapes is the current upstream for now

**Pull request closed without merging**